

# SENG 474
# Assignment 1 - Problem 4
# Nolan Kurylo
# V00893175
References:

1) https://pandas.pydata.org/pandas-docs/stable/reference/index.html

2) https://www.python-course.eu/Decision_Trees.php

3) https://medium.com/@lope.ai/decision-trees-from-scratch-using-id3-python-coding-it-up-6b79e3458de4

4) https://stackoverflow.com/

5) https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html


In [1]:
# Only categorical features
import pandas as pd
import numpy as np
import pprint
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.metrics import accuracy_score, zero_one_loss

np.random.seed(1337)

df = pd.read_csv('elections_clean.csv') # create dataframe of csv

label_vector = df.pop('Democrat') # this is the target/label vector

feature_list = ['Education','Religion','EthnicMale','EthnicFemale'] # categorical features 

df = df[feature_list] # only split on categorical features (same as question 2)
df = pd.get_dummies(df, feature_list) # one hot encoding of categorical features
df['Democrat'] = label_vector # place label vector at the end of df


training_set, validation_set, training_label, validation_label = train_test_split(df.drop('Democrat', axis=1), df['Democrat'], train_size=0.7) # split 70% traing, 30% validation
DecisionTree = tree.DecisionTreeClassifier(criterion='entropy') # generate decision tree
DecisionTree.fit(training_set, training_label) # train the tree using the 70% training set

validation_prediction = DecisionTree.predict(validation_set) #predict accuracy of tree using validation set
training_prediction = DecisionTree.predict(training_set) #predict accuracy of tree using training set

training_acc = accuracy_score(training_label, training_prediction) * 100.0 # validation set accuracy
training_err = zero_one_loss(training_label, training_prediction) * 100.0 # validation set error

validation_acc = accuracy_score(validation_label, validation_prediction) * 100.0 # validation set accuracy
validation_err = zero_one_loss(validation_label, validation_prediction) * 100.0 # validation set error

print("The training accuracy of the DecisionTree: " + str(training_acc) + "%")
print("The training error of the DecisionTree: " + str(training_err) + "%")
print()

print("The validation accuracy of the DecisionTree: " + str(validation_acc) + "%")
print("The validation error of the DecisionTree: " + str(validation_err) + "%")
print()

max_depth = DecisionTree.get_depth()
print("Maximum Depth of DecisionTree is: " + str(max_depth))




The training accuracy of the DecisionTree: 91.41299409359382%
The training error of the DecisionTree: 8.587005906406176%

The validation accuracy of the DecisionTree: 91.10169491525424%
The validation error of the DecisionTree: 8.898305084745761%

Maximum Depth of DecisionTree is: 8


# Tree Comparison 
In comparison to the Decision Tree built from scratch in Question 2, the structure and prediction accuracy of both tree are similar. When converting the
same categorical features used in Question 2 to be one-hot-encoded features with the scikit-learn tree, both trees have a a maximum depth near 8 and a prediction accuracy
around 91% and zero-one loss error of about 9%. They are similar since they use the same categorical features to split on and are both "entropy" based splitting algorithms, instead of using the gini coefficients. 

See above for code and results.